sqlmesh/core/table_diff.py

```diff
@@ -0,0 +1,195 @@
+    stats: t.Dict[str, float]
+    sample: pd.DataFrame
+
+    @property
+    def s_count(self) -> int:
+        """Count of the source."""
+        return int(self.stats["s_count"])
+
+    @property
+    def t_count(self) -> int:
+        """Count of the target."""
+        return int(self.stats["t_count"])
+
+    @property
+    def count_pct_change(self) -> float:
+        """The percentage change of the counts."""
+        return ((self.t_count - self.s_count) / self.s_count) * 100
+
+
+class TableDiff:
+    """Calculates differences between tables, taking into account schema and row level differences."""
+
+    def __init__(
+        self,
+        adapter: EngineAdapter,
+        source: TableName,
+        target: TableName,
+        on: t.List[str] | exp.Condition,
+        where: t.Optional[str | exp.Condition] = None,
+        dialect: DialectType = None,
+        limit: int = 20,
+    ):
+        self.adapter = adapter
+        self.source = source
+        self.target = target
+        self.where = exp.condition(where, dialect=dialect) if where else None
+        self.limit = limit
+
+        if isinstance(on, list):
+            self.on: exp.Condition = exp.and_(
+                *(
+                    exp.column(c, "s").eq(exp.column(c, "t"))
+                    | (exp.column(c, "s").is_(exp.null()) & exp.column(c, "t").is_(exp.null()))
+                    for c in on
+                )
+            )
+        else:
+            self.on = on
+
+        self._source_schema: t.Optional[t.Dict[str, exp.DataType]] = None
+        self._target_schema: t.Optional[t.Dict[str, exp.DataType]] = None
+        self._summary_diff: t.Optional[SummaryDiff] = None
+
+    @property
+    def source_schema(self) -> t.Dict[str, exp.DataType]:
+        if self._source_schema is None:
+            self._source_schema = self.adapter.columns(self.source)
+        return self._source_schema
+
+    @property
+    def target_schema(self) -> t.Dict[str, exp.DataType]:
+        if self._target_schema is None:
+            self._target_schema = self.adapter.columns(self.target)
+        return self._target_schema
+
+    def schema_diff(self) -> SchemaDiff:
+        return SchemaDiff(
+            source=self.source,
+            target=self.target,
+            source_schema=self.source_schema,
+            target_schema=self.target_schema,
+        )
+
+    def summary_diff(self) -> SummaryDiff:
+        if self._summary_diff is None:
+            s_selects = {c: exp.column(c, "s").as_(f"s__{c}") for c in self.source_schema}
+            t_selects = {c: exp.column(c, "t").as_(f"t__{c}") for c in self.target_schema}
+
+            s_index = []
+            t_index = []
+
+            for col in self.on.find_all(exp.Column):
+                col.not_().is_(exp.Null())
+                if col.table == "s":
+                    s_index.append(col)
+                elif col.table == "t":
+                    t_index.append(col)
+
+            comparisons = [
```

Comment:
It doesn't look like we report this anywhere.

Reply:
it's not reported, it's used for filtering